File: CPlusPlus/Add Two Numbers/AddTwoNumbers.cpp
//Definition for singly-linked list.
//struct ListNode 
//{
//    int val;
//   ListNode *next;
//    ListNode(int x) : val(x), next(NULL) {}
//};
 
class Solution 
{
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) 
    {
		ListNode *list1 = l1;
		ListNode *list2 = l2;
		ListNode l3(0);
        ListNode *ptr = &l3;
		bool carry = false;
		while (list1 || list2 || carry)
		{
			int sum = 0;
			sum = (list1 ? list1->val : 0) + (list2 ? list2->val : 0) + carry;
			if (sum > 9)
			{
				sum %= 10;
				carry = true;
			}
			else
			{
				carry = false;
			}
            
            ptr->next = new ListNode(sum);
            ptr = ptr->next; 
            if(list1)
                list1 = list1->next;
            if(list2)
			    list2 = list2->next;
		}
		return l3.next;

    }
};

File: Python/Add Two Numbers/addTwoNumbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        
        l3 = ListNode(0)
        
        walker = l3
        
        carry = 0
        
        total = 0
        
        while (l1 or l2 or carry):
        
            if l1:
                total += l1.val
                l1 = l1.next
        
            if l2:
                total += l2.val
                l2 = l2.next
                
            if carry:
                total += 1
                carry = 0
        
            if total > 9:
                carry = 1
                total %= 10
            
            walker.next = ListNode(total)
            
            total = 0
            
            walker = walker.next
        
        return l3.next
        
